db/sqldb/portfolio.go

```diff
@@ -22,18 +25,32 @@ func portfolioQueryStanza(where string) string {
 		portfolio.blob_id,
 		portfolio.admin_debug_enabled,
 		portfolio.number_of_rows,
-		ARRAY_AGG(portfolio_group_membership.portfolio_group_id),
-		ARRAY_AGG(portfolio_group_membership.created_at),
-		ARRAY_AGG(portfolio_initiative_membership.initiative_id),
-		ARRAY_AGG(portfolio_initiative_membership.added_by_user_id),
-		ARRAY_AGG(portfolio_initiative_membership.created_at)
+		portfolio_group_ids,
+		portfolio_group_created_ats,
+		initiative_ids,
+		initiative_added_by_user_ids,
+		initiative_created_ats
 	FROM portfolio
-	LEFT JOIN portfolio_group_membership 
-	ON portfolio_group_membership.portfolio_id = portfolio.id
-	LEFT JOIN portfolio_initiative_membership
-	ON portfolio_initiative_membership.portfolio_id = portfolio.id
-	%s
-	GROUP BY portfolio.id;`, where)
+	LEFT JOIN (
+		SELECT 
+			portfolio_id,
+			ARRAY_AGG(portfolio_group_id) as portfolio_group_ids,
+			ARRAY_AGG(created_at) as portfolio_group_created_ats
+		FROM portfolio_group_membership
+		WHERE portfolio_id IN (SELECT id FROM selected_portfolio_ids)
+		GROUP BY portfolio_id
+	) pgs ON pgs.portfolio_id = portfolio.id
+	LEFT JOIN (
+		SELECT
+			portfolio_id,
+			ARRAY_AGG(initiative_id) as initiative_ids,
+			ARRAY_AGG(added_by_user_id) as initiative_added_by_user_ids,
+			ARRAY_AGG(created_at) as initiative_created_ats
+		FROM portfolio_initiative_membership
+		WHERE portfolio_id IN (SELECT id FROM selected_portfolio_ids)
+		GROUP BY portfolio_id
+	) itvs ON itvs.portfolio_id = portfolio.id
+	%[1]s;`, where)
```

Comment:
I'm not sure if I understand these changes, specifically why these `LEFT JOIN`s are now structured as subqueries, since they just join on the portfolio ID anyway. I know you mentioned a "cardinality bug", but I don't follow how this fixes that, when the subqueries are grouping by the same thing we were initially grouping by anyway (portfolio ID).

Reply:
Great question. This wasn't clear to me and took a while to debug. Here's my best explanation for posterity:

- Joins happen before any aggregation or group by.
- 3-table joins will create rows based on the cartesian product of their rows. If we have a primary table (P) and two secondary tables (A) and (B), for a given primary key, if we have 3 values in A (A1, A2, A3), and 2 values in B (B1, B2), then the Cartesian product will have 6 values in the join.
- When we then group by the primary key, that row has six values for both A and B in the join result (A1, A3, A3, A1, A2, A3), (B1, B2, B1, B2, B1, B2).
- We could unwind this through uniqueness sets or similar. However, there are two complicating factors that make this much harder to do: (a) nulls (b) objects that are unbunded into different columns (c) (most difficult) objects that are unbundled into different columns across tables.

After playing around with some row-based solutions, I found this was the simplest way (and probably the most extensible/easy to change: do your group-bys where it's not over a join result, and then just select). This also has the advantage of not generating huge cross products when the cardinality of the subtables is large (as it might be in this case).

Reply:
>     * 3-table joins will create rows based on the cartesian product of their rows. If we have a primary table (P) and two secondary tables (A) and (B), for a given primary key, if we have 3 values in A (A1, A2, A3), and 2 values in B (B1, B2), then the Cartesian product will have 6 values in the join.

Huh yeah, sure enough:

```sql
CREATE TABLE p (
  id TEXT PRIMARY KEY NOT NULL
);

CREATE TABLE a (
  id TEXT PRIMARY KEY NOT NULL,
  p_id TEXT REFERENCES p (id) NOT NULL
);

CREATE TABLE b (
  id TEXT PRIMARY KEY NOT NULL,
  p_id TEXT REFERENCES p (id) NOT NULL
);


INSERT INTO p (id) VALUES ('p1');
INSERT INTO a (id, p_id) VALUES ('a1', 'p1'), ('a2', 'p1');
INSERT INTO b (id, p_id) VALUES ('b1', 'p1'), ('b2', 'p1'), ('b3', 'p1');

SELECT p.id, ARRAY_AGG(a.id), ARRAY_AGG(b.id)
FROM p
LEFT JOIN a ON p.id = a.p_id
LEFT JOIN b ON p.id = b.p_id
GROUP BY p.id;
```

produces

```
 id |      array_agg      |      array_agg
----+---------------------+---------------------
 p1 | {a2,a1,a2,a1,a2,a1} | {b1,b1,b2,b2,b3,b3}
```

When asking _the internet_ about this, I got the suggestion:

```sql
  SELECT p.id,
      (SELECT ARRAY_AGG(a.id) FROM a WHERE a.p_id = p.id),
      (SELECT ARRAY_AGG(b.id) FROM b WHERE b.p_id = p.id)
  FROM p;
```

Which involves subqueries but otherwise seems like the simplest approach, and I think applies here?

Reply:
Totally fair, that approach also would work. I think the downside of it is when we have multiple columns from `a` or `b` (and because memberships will often have an `id` and a `created at`, I think this is most cases), a nested query in the select statement then needs to be either (a) assumed order equivalent across multiple expressions, which might be the case or might not, or (b) needs to be multi-selected and the unnested (i.e. put into a composite object, then decomposed).